### PR TITLE
Implement matmul operator to Variable

### PR DESCRIPTION
**Pull request by okuta**
Rel #666.

### CODE REVIEW

File: chainer/functions/math/basic_math.py [hunk 1]
```diff
@@ -33,6 +34,14 @@ def _check_constant_type(value):
             'value must be scalar, ndarray, or Variable')
 
 
+def _check_constant_ndarray_type(value):
+    if isinstance(value, (numpy.ndarray, cuda.ndarray)):
```
Comment (delta2323) [inline chainer/functions/math/basic_math.py]: The following is simpler:
```
if not isinstance(...):
    raise ValueError(...)
```

File: tests/chainer_tests/functions_tests/math_tests/test_basic_math.py [hunk 2]
```diff
@@ -1,7 +1,8 @@
+import operator
+import sys
 import unittest
 
 import numpy
-import six
```
Comment (muupan) [inline tests/chainer_tests/functions_tests/math_tests/test_basic_math.py]: `six.assertRaisesRegex` is used in the code. Removing this line makes gpu tests fail.

**Comment by muupan** [comment]
LGTM! Though I think  this PR can be merged now, in the future it should be noted somewhere in the documents that operator `@` is now supported by `Variable`. I made an issue https://github.com/pfnet/chainer/issues/1382 for this.

File: chainer/functions/math/basic_math.py [hunk 3]
```diff
@@ -14,23 +15,17 @@ def _convert_value_to_string(value):
     if numpy.isscalar(value):
         if value < 0:
             return '({})'.format(value)
-        else:
```
Comment (unnonouno) [inline chainer/functions/math/basic_math.py]: Why are these fixes required?

File: tests/chainer_tests/functions_tests/math_tests/test_basic_math.py [hunk 4]
```diff
@@ -995,6 +997,71 @@ def test_gpu(self):
         self.check_backward(cuda.to_gpu(self.x), cuda.to_gpu(self.gy))
 
 
+@testing.parameterize(
+    {'left_const': False, 'right_const': False},
+    {'left_const': True, 'right_const': False},
+    {'left_const': False, 'right_const': True},
+)
+class TestMatMulVarVar(unittest.TestCase):
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/math_tests/test_basic_math.py]: Do you need to test other float types?

File: tests/chainer_tests/functions_tests/math_tests/test_basic_math.py [hunk 5]
```diff
@@ -995,6 +997,71 @@ def test_gpu(self):
...
+
+    def check_backward(self, x_data, y_data, z_grad):
+        if self.right_const:
+            op = lambda x: operator.matmul(x, y_data)
+            data = x_data,
+        elif self.left_const:
+            op = lambda y: operator.matmul(x_data, y)
+            data = y_data,
+        else:
+            op = lambda x, y: operator.matmul(x, y)
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/math_tests/test_basic_math.py]: `op = operator.matmul`?

File: tests/chainer_tests/functions_tests/math_tests/test_basic_math.py [hunk 6]
```diff
@@ -995,6 +997,71 @@ def test_gpu(self):
...
+            x = chainer.Variable(x_data)
+        if self.right_const:
+            y = y_data
+        else:
+            y = chainer.Variable(y_data)
+        z = operator.matmul(x, y)
+        gradient_check.assert_allclose(
+            operator.matmul(self.x, self.y), z.data, atol=1e-7, rtol=1e-7)
+
+    @unittest.skipUnless(sys.version_info[0] >= 3 and sys.version_info[1] >= 5,
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/math_tests/test_basic_math.py]: When python version is 4.0.0, this condition is skipped.

**Comment by okuta** [comment]
@unnonouno Please review this PR?

File: tests/chainer_tests/functions_tests/math_tests/test_basic_math.py [hunk 7]
```diff
@@ -995,6 +997,87 @@ def test_gpu(self):
...
+        if self.right_const:
+            y = y_data
+        else:
+            y = chainer.Variable(y_data)
+        z = operator.matmul(x, y)
+        if self.dtype == numpy.float16:
+            options = {'atol': 1e-3, 'rtol': 1e-3}
+        else:
+            options = {'atol': 1e-7, 'rtol': 1e-7}
+        gradient_check.assert_allclose(
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/math_tests/test_basic_math.py]: `testing.assert_allclose`?

File: tests/chainer_tests/functions_tests/math_tests/test_basic_math.py [hunk 8]
```diff
@@ -995,6 +997,87 @@ def test_gpu(self):
...
+            operator.matmul(self.x, self.y), z.data, **options)
+
+    @unittest.skipUnless(sys.version_info[0] >= 3 and sys.version_info[1] >= 5,
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/math_tests/test_basic_math.py]: `sys.version_info >= (3, 5)` ?

File: tests/chainer_tests/functions_tests/math_tests/test_basic_math.py [hunk 9]
```diff
@@ -995,6 +997,87 @@ def test_gpu(self):
...
+                         'Only for Python3.5 or higher')
+    @condition.retry(3)
+    def test_forward_cpu(self):
+        self.check_forward(self.x, self.y)
+
+    @unittest.skipUnless(sys.version_info[0] >= 3 and sys.version_info[1] >= 5,
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/math_tests/test_basic_math.py]: ditto

**Comment by unnonouno** [comment]
I wrote some comments.

File: tests/chainer_tests/functions_tests/math_tests/test_basic_math.py [hunk 10]
```diff
@@ -995,6 +997,87 @@ def test_gpu(self):
...
+            data = x_data, y_data
+
+        if self.dtype == numpy.float16:
+            options = {'eps': 1e-2, 'atol': 1e-2, 'rtol': 1e-1}
+        else:
+            options = {'atol': 1e-4, 'rtol': 1e-4}
+        gradient_check.check_backward(
+            op, data, z_grad, **options)
+
+    @unittest.skipUnless(sys.version_info[0] >= 3 and sys.version_info[1] >= 5,
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/math_tests/test_basic_math.py]: fix it

File: tests/chainer_tests/functions_tests/math_tests/test_basic_math.py [hunk 11]
```diff
@@ -995,6 +997,87 @@ def test_gpu(self):
...
+                         'Only for Python3.5 or higher')
+    def test_backward_cpu(self):
+        self.check_backward(self.x, self.y, self.gz)
+
+    @attr.gpu
+    @unittest.skipUnless(
+        (sys.version_info.major >= 4 or
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/math_tests/test_basic_math.py]: fix it

File: chainer/functions/math/matmul.py [hunk 12]
```diff
@@ -129,8 +129,8 @@ def check_type_forward(self, in_types):
         a_type, b_type = in_types
 
         type_check.expect(
-            a_type.dtype == numpy.float32,
-            b_type.dtype == numpy.float32
+            a_type.dtype.kind == 'f',
+            a_type.dtype == b_type.dtype
```
Comment (unnonouno) [inline chainer/functions/math/matmul.py]: Do we need to fix test_matmul?

Comment (okuta) [inline chainer/functions/math/matmul.py]: I added tests for float16 and float64.

File: chainer/functions/math/basic_math.py [hunk 13]
```diff
@@ -431,6 +432,97 @@ def rpow(lhs, rhs):  # rhs ** lhs
...
+
+
+def matmul(lhs, rhs):  # lhs @ rhs
+    if isinstance(rhs, variable.Variable):
+        return MatMulVarVar()(lhs, rhs)
+    _check_constant_type(rhs)
+    return MatMulVarConst(rhs)(lhs)
+
+
+def rmatmul(lhs, rhs):  # rhs / lhs
```
Comment (unnonouno) [inline chainer/functions/math/basic_math.py]: `rhs @ lhs` ?

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**